### PR TITLE
upgrade to NAN v2

### PR DESCRIPTION
**Pull request by julianduque**
Getting ready for Node.js v4.0

### CODE REVIEW

**Comment by ChALkeR** [comment]
Adding to the list: https://github.com/nodejs/node/issues/2798.

File: src/serialport_poller.cpp [hunk 1]
```diff
@@ -50,55 +50,52 @@ void SerialportPoller::callCallback(int status) {
...
-  tpl->PrototypeTemplate()->Set(NanNew<String>("close"),
-      NanNew<FunctionTemplate>(Close)->GetFunction());
+  tpl->PrototypeTemplate()->Set(Nan::New<String>("close").ToLocalChecked(),
+      Nan::New<FunctionTemplate>(Close)->GetFunction());
 
   // SerialportPoller.start()
-  tpl->PrototypeTemplate()->Set(NanNew<String>("start"),
-      NanNew<FunctionTemplate>(Start)->GetFunction());
-
-  NanAssignPersistent<FunctionTemplate>(serialportpoller_constructor, tpl);
```
Comment (jacobrosenthal) [inline src/serialport_poller.cpp]: no reset?

File: src/serialport.cpp [hunk 2]
```diff
@@ -256,39 +256,38 @@ void EIO_AfterUpdate(uv_work_t* req) {
...
+    Nan::ThrowTypeError("Third argument must be a function");
+    return;
   }
-  v8::Local<v8::Function> callback = args[2].As<v8::Function>();
+  v8::Local<v8::Function> callback = info[2].As<v8::Function>();
 
   WriteBaton* baton = new WriteBaton();
   memset(baton, 0, sizeof(WriteBaton));
   baton->fd = fd;
-  NanAssignPersistent<v8::Object>(baton->buffer, buffer);
```
Comment (jacobrosenthal) [inline src/serialport.cpp]: @ghostoy does a baton->buffer.Reset(buffer); here as a replacement..

**Comment by jacobrosenthal** [comment]
This doesnt build for me on osx yosemite 0.10.38
```
  CXX(target) Release/obj.target/serialport/src/serialport.o
../src/serialport.cpp:363:16: error: no member named 'Reset' in 'v8::Persistent<v8::Object>'
  data->buffer.Reset();
  ~~~~~~~~~~~~ ^
../src/serialport.cpp:417:30: error: no member named 'Reset' in 'v8::Persistent<v8::Object>'
        del_q->baton->buffer.Reset();
        ~~~~~~~~~~~~~~~~~~~~ ^
../src/serialport.cpp:655:30: error: no matching constructor for initialization of 'Nan::Utf8String'
  Nan::Utf8String *str = new Nan::Utf8String(v8str);
                             ^               ~~~~~
../node_modules/nan/nan.h:1174:23: note: candidate constructor not viable: no known conversion from 'const v8::Handle<v8::String>' to
      'v8::Local<v8::Value>' for 1st argument
  NAN_INLINE explicit Utf8String(v8::Local<v8::Value> from) :
                      ^
../node_modules/nan/nan.h:1207:33: note: candidate constructor not viable: no known conversion from 'const v8::Handle<v8::String>' to
      'const Nan::Utf8String' for 1st argument
  NAN_DISALLOW_ASSIGN_COPY_MOVE(Utf8String)
                                ^
../node_modules/nan/nan.h:109:23: note: expanded from macro 'NAN_DISALLOW_ASSIGN_COPY_MOVE'
    NAN_DISALLOW_COPY(CLASS)                                                   \
                      ^
../node_modules/nan/nan.h:91:35: note: expanded from macro 'NAN_DISALLOW_COPY'
# define NAN_DISALLOW_COPY(CLASS) CLASS(const CLASS&);
```

File: src/serialport.cpp (hunk 2 above)
Comment (julianduque) [inline src/serialport.cpp]: @jacobrosenthal Reset here isn't working:

```
 CXX(target) Release/obj.target/serialport/src/serialport.o
../src/serialport.cpp:287:17: error: no matching member function for call to 'Reset'
  baton->buffer.Reset(buffer);
  ~~~~~~~~~~~~~~^~~~~
/Users/julianduque/.node-gyp/4.0.0/include/node/v8.h:500:18: note: candidate function template not viable: requires 2 arguments, but 1 was provided
  V8_INLINE void Reset(Isolate* isolate, const Local<S>& other);
                 ^
/Users/julianduque/.node-gyp/4.0.0/include/node/v8.h:507:18: note: candidate function template not viable: requires 2 arguments, but 1 was provided
  V8_INLINE void Reset(Isolate* isolate, const PersistentBase<S>& other);
                 ^
/Users/julianduque/.node-gyp/4.0.0/include/node/v8.h:494:18: note: candidate function not viable: requires 0 arguments, but 1 was provided
  V8_INLINE void Reset();
                 ^
1 error generated.
```

**Comment by fivdi** [comment]
I performed some tests with this PR on the BeagleBone Black with two different test setups.

1) Debian Wheezy, gcc/g++ 4.6, Node.js v0.10.38
   Result: compile errors on install (see below)

2) Debian Jessie, gcc/g++ 4.9, Node.js v4.1.1
   Result: installed successfully and a simple test ran successfully
```
root@beaglebone:~/dev/pr571# node -v
v0.10.38
root@beaglebone:~/dev/pr571# uname -a
Linux beaglebone 3.8.13-bone70 #1 SMP Fri Jan 23 02:15:42 UTC 2015 armv7l GNU/Linux
root@beaglebone:~/dev/pr571# npm install "git://github.com/julianduque/node-serialport.git#master"
-
> serialport@1.7.4 install /root/dev/pr571/node_modules/serialport
> node-pre-gyp install --fallback-to-build

make: Entering directory `/root/dev/pr571/node_modules/serialport/build'
  CXX(target) Release/obj.target/serialport/src/serialport.o
../src/serialport.cpp: In function ‘void EIO_AfterWrite(uv_work_t*)’:
../src/serialport.cpp:363:16: error: ‘class v8::Persistent<v8::Object>’ has no member named ‘Reset’
../src/serialport.cpp: In function ‘void EIO_AfterClose(uv_work_t*)’:
../src/serialport.cpp:417:30: error: ‘class v8::Persistent<v8::Object>’ has no member named ‘Reset’
../src/serialport.cpp: In function ‘SerialPortParity ToParityEnum(const v8::Handle<v8::String>&)’:
../src/serialport.cpp:655:51: error: no matching function for call to ‘Nan::Utf8String::Utf8String(const v8::Handle<v8::String>&)’
../src/serialport.cpp:655:51: note: candidates are:
../node_modules/nan/nan.h:1207:3: note: Nan::Utf8String::Utf8String(const Nan::Utf8String&)
../node_modules/nan/nan.h:1207:3: note:   no known conversion for argument 1 from ‘const v8::Handle<v8::String>’ to ‘const Nan::Utf8String&’
../node_modules/nan/nan.h:1174:23: note: Nan::Utf8String::Utf8String(v8::Local<v8::Value>)
../node_modules/nan/nan.h:1174:23: note:   no known conversion for argument 1 from ‘const v8::Handle<v8::String>’ to ‘v8::Local<v8::Value>’
make: *** [Release/obj.target/serialport/src/serialport.o] Error 1
make: Leaving directory `/root/dev/pr571/node_modules/serialport/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/share/node-gyp/lib/build.js:267:23)
gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:820:12)
gyp ERR! System Linux 3.8.13-bone70
gyp ERR! command "nodejs" "/usr/bin/node-gyp" "build" "--fallback-to-build" "--module=/root/dev/pr571/node_modules/serialport/build/serialport/v1.7.4/Release/node-v11-linux-arm/serialport.node" "--module_name=serialport" "--module_path=/root/dev/pr571/node_modules/serialport/build/serialport/v1.7.4/Release/node-v11-linux-arm"
gyp ERR! cwd /root/dev/pr571/node_modules/serialport
gyp ERR! node -v v0.10.38
gyp ERR! node-gyp -v v0.12.2
gyp ERR! not ok 
node-pre-gyp ERR! build error 
node-pre-gyp ERR! stack Error: Failed to execute 'node-gyp build --fallback-to-build --module=/root/dev/pr571/node_modules/serialport/build/serialport/v1.7.4/Release/node-v11-linux-arm/serialport.node --module_name=serialport --module_path=/root/dev/pr571/node_modules/serialport/build/serialport/v1.7.4/Release/node-v11-linux-arm' (1)
node-pre-gyp ERR! stack     at ChildProcess.<anonymous> (/root/dev/pr571/node_modules/serialport/node_modules/node-pre-gyp/lib/util/compile.js:83:29)
node-pre-gyp ERR! stack     at ChildProcess.emit (events.js:98:17)
node-pre-gyp ERR! stack     at maybeClose (child_process.js:766:16)
node-pre-gyp ERR! stack     at Process.ChildProcess._handle.onexit (child_process.js:833:5)
node-pre-gyp ERR! System Linux 3.8.13-bone70
node-pre-gyp ERR! command "node" "/root/dev/pr571/node_modules/serialport/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
node-pre-gyp ERR! cwd /root/dev/pr571/node_modules/serialport
node-pre-gyp ERR! node -v v0.10.38
node-pre-gyp ERR! node-pre-gyp -v v0.6.11
node-pre-gyp ERR! not ok 
Failed to execute 'node-gyp build --fallback-to-build --module=/root/dev/pr571/node_modules/serialport/build/serialport/v1.7.4/Release/node-v11-linux-arm/serialport.node --module_name=serialport --module_path=/root/dev/pr571/node_modules/serialport/build/serialport/v1.7.4/Release/node-v11-linux-arm' (1)
npm WARN This failure might be due to the use of legacy binary "node"
npm WARN For further explanations, please read
/usr/share/doc/nodejs/README.Debian
 
npm ERR! serialport@1.7.4 install: `node-pre-gyp install --fallback-to-build`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the serialport@1.7.4 install script.
npm ERR! This is most likely a problem with the serialport package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-pre-gyp install --fallback-to-build
npm ERR! You can get their info via:
npm ERR!     npm owner ls serialport
npm ERR! There is likely additional logging output above.

npm ERR! System Linux 3.8.13-bone70
npm ERR! command "/usr/bin/nodejs" "/usr/bin/npm" "install" "git://github.com/julianduque/node-serialport.git#master"
npm ERR! cwd /root/dev/pr571
npm ERR! node -v v0.10.38
npm ERR! npm -v 1.4.21
npm ERR! code ELIFECYCLE
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /root/dev/pr571/npm-debug.log
npm ERR! not ok code 0
root@beaglebone:~/dev/pr571# 
```

**Comment by jacobrosenthal** [comment]
Thanks for PR @julianduque, we ended up merging #566 as it got more work and testing for whatever reasons

**Comment by julianduque** [comment]
@jacobrosenthal perfect :clap:

**Pull request closed without merging**